inconsistent identifier capitalization between protoc-gen-go and protoc-gen-grpc-gateway

In the following proto file, notice the underscores, numbers (and possibly other characters) in identifiers:
```
syntax = "proto3";
import "google/api/annotations.proto";
package template;

service Greeter {
  rpc Sen_dGet (Temp_lateRequest) returns (template_response) {
      option (google.api.http).get = "/get";
  }
  rpc SendGe2t (Temp_lateRequest) returns (template_response) {
      option (google.api.http).get = "/get";
  }
}

message Temp_lateRequest {
  string name = 1;
}

message template_response {
  string name = 1;
}
```
After compiling such file, building the executables produces these errors:
```
../template/template.pb.gw.go:36:15: undefined: Temp_lateRequest
../template/template.pb.gw.go:43:20: client.Sen_dGet undefined (type GreeterClient has no field or method Sen_dGet)
../template/template.pb.gw.go:53:15: undefined: Temp_lateRequest
../template/template.pb.gw.go:60:20: client.SendGe2t undefined (type GreeterClient has no field or method SendGe2t, but does have SendGe2T)
```

Looking at the pb.go file, we can see the identifiers are different from those in pb.gw.go file
```
type GreeterClient interface {
	SenDGet(ctx context.Context, in *TempLateRequest, opts ...grpc.CallOption) (*TemplateResponse, error)
	SendGe2T(ctx context.Context, in *TempLateRequest, opts ...grpc.CallOption) (*TemplateResponse, error)
}
```
Judging from the code, this happens because protoc-gen-go uses CamelCase
```
./grpc/grpc.go:159:	servName := generator.CamelCase(origServName)
./grpc/grpc.go:283:	methName := generator.CamelCase(origMethName)
```

whereas protoc-gen-grpc-gateway simply calls strings.Title
```
./protoc-gen-grpc-gateway/gengateway/template.go:85:		svcName := strings.Title(*svc.Name)
./protoc-gen-grpc-gateway/gengateway/template.go:89:			methName := strings.Title(*meth.Name)
```

UPDATE:
@johanbrandhorst Steps I followed to reproduce your issue.

Copied the proto file and compiled it with the following command.
```sh
$ protoc -I/usr/local/include -I. \
-I$GOPATH/src \
-I$GOPATH/src/github.com/grpc-ecosystem/grpc-gateway/third_party/googleapis \
--plugin=protoc-gen-grpc-gateway=$GOPATH/bin/protoc-gen-grpc-gateway \
--grpc-gateway_out=logtostderr=true:.  \
$GOPATH/src/echo/echo.proto
```
Upon running the above command a echo.pb.gw.go file wasgenerated.

After creating an empty project, I ran `go install project` after which I got a similar error which said.

```sh
#echo/echo 
echo/echo.pb.gw.go:36:15: undefined: Temp_lateRequest
echo/echo.pb.gw.go:53:15: undefined: Temp_lateRequest
```
To test this further I removed the underscores in `Temp_lateRequest` in the .proto file giving me `TemplateRequest`. After compiling this new proto file and running `go install project` I got the following error.

```sh
# echo/echo
echo/echo.pb.gw.go:36:15: undefined: TemplateRequest
echo/echo.pb.gw.go:53:15: undefined: TemplateRequest
```
After the above error, all underscores were removed from the echo.proto file to get the following file which was compiled.

```proto
syntax = "proto3";

import "google/api/annotations.proto";

package echo;

service Greeter {
  rpc SendGet (TemplateRequest) returns (templateresponse) {
      option (google.api.http).get = "/get";
  }
  rpc SendGe2t (TemplateRequest) returns (templateresponse) {
      option (google.api.http).get = "/get";
  }
}

message TemplateRequest {
  string name = 1;
}

message templateresponse {
  string name = 1;
}
```

And upon running `go install project` I got the following errors.
```sh
# echo/echo
echo/echo.pb.gw.go:36:15: undefined: TemplateRequest
echo/echo.pb.gw.go:43:20: client.SendGet undefined (type GreeterClient has no field or method SendGet, but does have SenDGet)
echo/echo.pb.gw.go:53:15: undefined: TemplateRequest
```

This seems to happen only during compilation of the app.

--- a/protoc-gen-grpc-gateway/gengateway/template.go
+++ b/protoc-gen-grpc-gateway/gengateway/template.go
@@ -112,7 +112,6 @@ func (b binding) FieldMaskField() string {
 			fieldMaskField = f
 		}
 	}
-
 	if fieldMaskField != nil {
 		return generator2.CamelCase(fieldMaskField.GetName())
 	}
@@ -145,13 +144,18 @@ func applyTemplate(p param, reg *descriptor.Registry) (string, error) {
 		return "", err
 	}
 	var targetServices []*descriptor.Service
+
+	for _, msg := range p.Messages {
+		msgName := generator2.CamelCase(*msg.Name)
+		msg.Name = &msgName
+	}
 	for _, svc := range p.Services {
 		var methodWithBindingsSeen bool
-		svcName := strings.Title(*svc.Name)
+		svcName := generator2.CamelCase(*svc.Name)
 		svc.Name = &svcName
 		for _, meth := range svc.Methods {
 			glog.V(2).Infof("Processing %s.%s", svc.GetName(), meth.GetName())
-			methName := strings.Title(*meth.Name)
+			methName := generator2.CamelCase(*meth.Name)
 			meth.Name = &methName
 			for _, b := range meth.Bindings {
 				methodWithBindingsSeen = true

--- a/protoc-gen-grpc-gateway/gengateway/template_test.go
+++ b/protoc-gen-grpc-gateway/gengateway/template_test.go
@@ -481,3 +481,98 @@ func TestAllowPatchFeature(t *testing.T) {
 		}
 	}
 }
+
+func TestIdentifierCapitalization(t *testing.T){
+	msgdesc1 := &protodescriptor.DescriptorProto{
+		Name: proto.String("Exam_pleRequest"),
+	}
+	msgdesc2 := &protodescriptor.DescriptorProto{
+		Name: proto.String("example_response"),
+	}
+	meth1 := &protodescriptor.MethodDescriptorProto{
+		Name:       proto.String("ExampleGe2t"),
+		InputType:  proto.String("Exam_pleRequest"),
+		OutputType: proto.String("example_response"),	
+	}
+	meth2 := &protodescriptor.MethodDescriptorProto{
+		Name:       proto.String("Exampl_eGet"),
+		InputType:  proto.String("Exam_pleRequest"),
+		OutputType: proto.String("example_response"),	
+	}
+	svc := &protodescriptor.ServiceDescriptorProto{
+		Name:   proto.String("Example"),
+		Method: []*protodescriptor.MethodDescriptorProto{meth1, meth2},
+	}
+	msg1 := &descriptor.Message{
+		DescriptorProto: msgdesc1,
+	}
+	msg2 := &descriptor.Message{
+		DescriptorProto: msgdesc2,
+	}
+	file := descriptor.File{
+		FileDescriptorProto: &protodescriptor.FileDescriptorProto{
+			Name:        proto.String("example.proto"),
+			Package:     proto.String("example"),
+			Dependency:  []string{"a.example/b/c.proto", "a.example/d/e.proto"},
+			MessageType: []*protodescriptor.DescriptorProto{msgdesc1, msgdesc2},
+			Service:     []*protodescriptor.ServiceDescriptorProto{svc},
+		},
+		GoPkg: descriptor.GoPackage{
+			Path: "example.com/path/to/example/example.pb",
+			Name: "example_pb",
+		},
+		Messages: []*descriptor.Message{msg1, msg2},
+		Services: []*descriptor.Service{
+			{
+				ServiceDescriptorProto: svc,
+				Methods: []*descriptor.Method{
+					{
+						MethodDescriptorProto: meth1,
+						RequestType:           msg1,
+						ResponseType:          msg1,
+						Bindings: []*descriptor.Binding{
+							{
+								HTTPMethod: "GET",
+								Body:       &descriptor.Body{FieldPath: nil},
+							},
+						},
+					},
+				},
+			},
+			{
+				ServiceDescriptorProto: svc,
+				Methods: []*descriptor.Method{
+					{
+						MethodDescriptorProto: meth2,
+						RequestType:           msg2,
+						ResponseType:          msg2,
+						Bindings: []*descriptor.Binding{
+							{
+								HTTPMethod: "GET",
+								Body:       &descriptor.Body{FieldPath: nil},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	got, err := applyTemplate(param{File: crossLinkFixture(&file), RegisterFuncSuffix: "Handler", AllowPatchFeature: true}, descriptor.NewRegistry())
+	if err != nil {
+		t.Errorf("applyTemplate(%#v) failed with %v; want success", file, err)
+		return
+	}
+	if want := `msg, err := client.ExampleGe2T(ctx, &protoReq, grpc.Header(&metadata.HeaderMD)`; !strings.Contains(got, want) {
+		t.Errorf("applyTemplate(%#v) = %s; want to contain %s", file, got, want)
+	}
+	if want := `msg, err := client.ExamplEGet(ctx, &protoReq, grpc.Header(&metadata.HeaderMD)`; !strings.Contains(got, want) {
+		t.Errorf("applyTemplate(%#v) = %s; want to contain %s", file, got, want)
+	}
+	if want := `var protoReq ExamPleRequest`; !strings.Contains(got, want) {
+		t.Errorf("applyTemplate(%#v) = %s; want to contain %s", file, got, want)
+	}
+	if want := `var protoReq ExampleResponse`; !strings.Contains(got, want) {
+		t.Errorf("applyTemplate(%#v) = %s; want to contain %s", file, got, want)
+	}
+}

